### PR TITLE
feat(dflash2): co-train the DFlash2 drafter against SGLang rollout

### PR DESCRIPTION
**Pull request by khazic**
## What does this PR do?

Adds online co-training for the DFlash2 drafter against SGLang rollout, completing the engine matrix started in #77 (vLLM). Stacked on #77: the first six commits here are that PR, and this one builds on its DFlash2 checkpoint-contract helpers; only the commits from `feat(dflash2): co-train the DFlash2 drafter against SGLang rollout` onward are new.

SGLang serves DFlash2 through its DFLASH speculative worker, dispatching the DFlash2 draft (dynamic convolutions plus candidate selector) from the checkpoint's `DFlash2DraftModel` architecture and building the modules from `dflash_config`. That support lives on sglang main only; no tagged release up to 0.5.18 ships it.

### Changes

- **Server-args mapping**: `DFLASH2` maps to SGLang `speculative_algorithm=DFLASH`, with the same fail-loud contracts as the vLLM path (capability probe for the `DFlash2DraftModel` class, checkpoint architecture and `dflash_config` hyperparameter validation), reusing #77's engine-agnostic validators.
- **Block-size contract**: SGLang uses `speculative_num_draft_tokens` directly as the DFlash block size (`--speculative-dflash-block-size` is an alias of it), unlike vLLM's bonus-token-plus-tokens, so the check here pins `spec_verify_tokens == dflash2_block_size` (8 by default) instead of `block_size - 1`.
- **Aux-hidden collection gate**: DFLASH2 joins DFLASH/DSPARK in `_drafter_uses_dflash_aux_hidden` (same concatenated target-layer layout). Note sglang main currently rejects `return_hidden_states` for the DFLASH worker, so the example collects hidden states from the trainer's old-logprob pass instead of `collect_hidden_states_from_sgl`.
- **Weight publish**: the selector codebooks are renamed to the engine spelling (`..._codebook.weight` to the bare parameter) before bucketing; the SGLang publish path forwards trainer names verbatim otherwise, and SGLang's `load_weights` silently drops unresolved names.
- **Draft-model detection fix (affects plain DFLASH too)**: `_is_sglang_eagle_draft_model` now recognizes the DFlash family. SGLang's DFlash draft classes spell neither "eagle" nor `draft_vocab_size`, so the custom draft weight loader silently dropped every published DFlash-family tensor while the target loader loaded into the draft.
- **sglang-main compatibility in the weight route**: `LocalSerializedTensor` moved to `model_runner_components.weight_updater` (resolved once, with an import fallback); `UpdateWeightsFromTensorReqInput` became a fixed msgspec struct, so the route probes the request's fields once, sets only what exists, and validates the split-routing contract on every rank before the gather (a rank-0-only raise would leave the other ranks hung in the collective) with the SPECO custom weight loaders as the fallback mechanism, failing loud when neither exists.
- **Overlay-level rejection of `collect_hidden_states_from_sgl` for DFLASH2**: sglang refuses `return_hidden_states` for the DFLASH worker, so the server-args builder rejects the combination with the old-logprob fix in the message instead of letting engine startup die deep inside `dflash_utils`.
- The `vllm_runtime` DFlash2 helpers are imported lazily at the DFlash2 call sites, so a pure-SGLang process never pays that module's import-time patches; the shared checkpoint validator now words its errors for the engine that raised them.
- **Allocator guard follow-up to #77**: the SGLang draft publish stages and serializes its IPC tensors inside the same non-expandable-segments guard as the vLLM send; and since torch has no getter for the prior state, the guard now restores expandable segments only when `PYTORCH_CUDA_ALLOC_CONF` explicitly requests them. Restoring `True` unconditionally poisoned later allocations in runs that never enabled them, and the initial target sync then failed with `pidfd_getfd` on kernels older than Linux 5.6.
- Example `examples/run_qwen3-8b_drafter_dflash2_sglang.sh`, a `gpu/sglang/dflash2` arm in `ci/run_example_test.sh` (with the per-engine `spec_verify_tokens` default), README rows and prose.

## Test

CPU: `tests/integration/test_sglang_dflash2_runtime_contract.py` (new: server-args mapping, capability probe against a real and a faked sglang, checkpoint rejection, block-size contract, aux-hidden gate, draft-model detection) plus the updated DFlash2/allocator/adapter/vLLM contract tests: 74 passed in both validation environments (one with sglang main installed, one without).

GPU (single node, sglang main + torch 2.13, `disable_cuda_graph` for the smoke): GRPO on GSM8K, Qwen3-4B target, the converted `mgoin/Qwen3-4B-speculator.dflash2` drafter, via `ci/run_example_test.sh gpu sglang dflash2`, 8 steps:

- 4 collect / train / publish cycles completed (`drafter/trained=1`, `drafter/published=1` on steps 2/4/6/8), 80 drafter optimizer steps, per-cycle loss 3.80 -> 3.25, 3.52 -> 3.00, ... -> 2.41, 2.46;
- each publish hot-loads the full DFlash2 draft (convolutions and selector codebooks included) through the custom draft weight loader in ~0.6 s per rank (3 buckets at 512 MB);
- SGLang's decode logs report the DFlash2 draft speculating end to end with the selector path active: `accept len` 3.94 - 4.58 (accept rate 0.42 - 0.51) across runs;
- exit code 0 (the smoke was re-run after the review cleanups with the same result).

Full sanitized log attached in the comments.

## Notes

- Requires sglang main; the capability probe fails loud on older builds with a clear message. vLLM co-training is unaffected (#77 covers it and its CI is green).
- `collect_hidden_states_from_sgl` for the DFlash family is blocked by sglang main itself (`DFLASH speculative decoding does not support return_hidden_states yet`); when sglang lifts that, the aux-hidden gate here already admits DFLASH2.

### CODE REVIEW

**Comment by khazic** [comment]
Sanitized log of the SGLang GPU validation run (internal paths and hosts rewritten). The excerpt keeps the launch command, every per-step metrics line, the drafter train/publish lifecycle, a sample of SGLang's own `accept len` decode lines, and the exit marker; the complete file is on the fork's log branch:

- [sgl_dflash2_cotrain_smoke_8steps.log](https://github.com/khazic/verl-SpeCo-lao/blob/logs/dflash2-vllm-cotrain-validation/sgl_dflash2_cotrain_smoke_8steps.log)

<details><summary>SGLang DFlash2 co-train smoke, 8 GRPO steps (excerpt)</summary>

```text
1d0c5c5 fix(sglang): address the code-review findings on the DFlash2 co-train path
sglang 0.5.19 verl 0.8.0.dev
+ project_name=verl_grpo_example_dflash2_drafter
+ python3 -m verl_speco.main algorithm.adv_estimator=grpo ray_kwargs.ray_init.num_cpus=64 +ray_kwargs.ray_init._system_config.prestart_worker_first_driver=false +ray_kwargs.ray_init._system_config.num_workers_soft_limit=8 data.train_files=/ ...
(SpecoSGLangHttpServer pid=729143) WARNING:2026-09-02 18:08:23,930:Installed SPECO SGLang runtime: algorithm=DFLASH2 draft_model=/data/models/Qwen3-4B-DFlash2
(SpecoSGLangHttpServer pid=729151) WARNING:2026-09-02 18:08:23,930:Installed SPECO SGLang runtime: algorithm=DFLASH2 draft_model=/data/models/Qwen3-4B-DFlash2 [repeated 7x across cluster]
(SpecoSGLangHttpServer pid=729151) [2026-09-02 18:11:20] Decode batch, #running-req: 8, #token: 2133, token usage: 0.01, accept len: 3.89, accept rate: 0.41, cuda graph: False, gen throughput (token/s): 13.34, #queue-req ...
step:1 - drafter/collected_samples:np.float64(0.0) - drafter/trained:np.float64(0.0) - drafter/published:np.float64(0.0) - critic/score/mean:0.171875 - response_length/mean:546.734375 - timing_s/gen:44.542254246771336
(SpecoSGLangHttpServer pid=729146) [2026-09-02 18:12:21] Decode batch, #running-req: 1, #token: 1011, token usage: 0.01, accept len: 4.42, accept rate: 0.49, cuda graph: False, gen throughput (token/s): 73.46, #queue-req ...
(SpecoWorker pid=731472) WARNING:2026-09-02 18:13:01,558:[drafter loss] step=1 optimizer_step_total=1 lr=1.000e-05 loss=3.7941 vloss=0.0000 ploss=3.7941
(SpecoWorker pid=731468) WARNING:2026-09-02 18:13:07,865:[drafter loss] step=20 optimizer_step_total=20 lr=1.000e-05 loss=3.3479 vloss=0.0000 ploss=3.3479 [repeated 32x across cluster]
(WorkerDict pid=728173) WARNING:2026-09-02 18:13:24,440:[speco sglang draft update] done global_steps=2 buckets=3 bucket_mb=512 elapsed_sec=0.648
step:2 - drafter/collected_samples:np.float64(8.0) - drafter/trained:np.float64(1.0) - drafter/published:np.float64(1.0) - critic/score/mean:0.5 - response_length/mean:543.28125 - timing_s/gen:22.33904842287302
(WorkerDict pid=728175) WARNING:2026-09-02 18:13:24,660:[speco sglang draft update] done global_steps=2 buckets=3 bucket_mb=512 elapsed_sec=0.689 [repeated 7x across cluster]
step:3 - drafter/collected_samples:np.float64(0.0) - drafter/trained:np.float64(0.0) - drafter/published:np.float64(0.0) - critic/score/mean:0.6875 - response_length/mean:588.28125 - timing_s/gen:22.58012445271015
(SpecoSGLangHttpServer pid=729143) [2026-09-02 18:14:03] Decode batch, #running-req: 8, #token: 2047, token usage: 0.01, accept len: 3.77, accept rate: 0.40, cuda graph: False, gen throughput (token/s): 51.15, #queue-req ...
(SpecoWorker pid=731469) WARNING:2026-09-02 18:14:35,195:[drafter loss] step=1 optimizer_step_total=21 lr=1.000e-05 loss=3.2458 vloss=0.0000 ploss=3.2458
(SpecoWorker pid=731468) WARNING:2026-09-02 18:14:41,477:[drafter loss] step=20 optimizer_step_total=40 lr=1.000e-05 loss=2.5180 vloss=0.0000 ploss=2.5180 [repeated 31x across cluster]
(WorkerDict pid=728171) WARNING:2026-09-02 18:14:54,299:[speco sglang draft update] done global_steps=4 buckets=3 bucket_mb=512 elapsed_sec=0.633
step:4 - drafter/collected_samples:np.float64(8.0) - drafter/trained:np.float64(1.0) - drafter/published:np.float64(1.0) - critic/score/mean:0.796875 - response_length/mean:438.09375 - timing_s/gen:16.41484574228525
(WorkerDict pid=728175) WARNING:2026-09-02 18:14:54,422:[speco sglang draft update] done global_steps=4 buckets=3 bucket_mb=512 elapsed_sec=0.683 [repeated 7x across cluster]
(SpecoSGLangHttpServer pid=729142) [2026-09-02 18:15:09] Decode batch, #running-req: 1, #token: 1170, token usage: 0.01, accept len: 4.55, accept rate: 0.51, cuda graph: False, gen throughput (token/s): 74.51, #queue-req ...
step:5 - drafter/collected_samples:np.float64(0.0) - drafter/trained:np.float64(0.0) - drafter/published:np.float64(0.0) - critic/score/mean:0.90625 - response_length/mean:497.203125 - timing_s/gen:19.75864229351282
(SpecoWorker pid=731469) WARNING:2026-09-02 18:16:08,630:[drafter loss] step=1 optimizer_step_total=41 lr=1.000e-05 loss=3.3699 vloss=0.0000 ploss=3.3699
(SpecoWorker pid=731468) WARNING:2026-09-02 18:16:14,885:[drafter loss] step=20 optimizer_step_total=60 lr=1.000e-05 loss=3.1021 vloss=0.0000 ploss=3.1021 [repeated 32x across cluster]
(WorkerDict pid=728176) WARNING:2026-09-02 18:16:27,733:[speco sglang draft update] done global_steps=6 buckets=3 bucket_mb=512 elapsed_sec=0.619
step:6 - drafter/collected_samples:np.float64(8.0) - drafter/trained:np.float64(1.0) - drafter/published:np.float64(1.0) - critic/score/mean:0.8125 - response_length/mean:570.421875 - timing_s/gen:22.290636375546455
(WorkerDict pid=728172) WARNING:2026-09-02 18:16:27,859:[speco sglang draft update] done global_steps=6 buckets=3 bucket_mb=512 elapsed_sec=0.670 [repeated 7x across cluster]
(SpecoSGLangHttpServer pid=729142) [2026-09-02 18:16:35] Decode batch, #running-req: 2, #token: 1219, token usage: 0.01, accept len: 4.39, accept rate: 0.48, cuda graph: False, gen throughput (token/s): 318.65, #queue-re ...
step:7 - drafter/collected_samples:np.float64(0.0) - drafter/trained:np.float64(0.0) - drafter/published:np.float64(0.0) - critic/score/mean:0.859375 - response_length/mean:517.515625 - timing_s/gen:20.78101073205471
(SpecoSGLangHttpServer pid=729150) [2026-09-02 18:17:18] Decode batch, #running-req: 1, #token: 1265, token usage: 0.01, accept len: 4.54, accept rate: 0.51, cuda graph: False, gen throughput (token/s): 78.24, #queue-req ...
(SpecoWorker pid=731469) WARNING:2026-09-02 18:17:42,053:[drafter loss] step=1 optimizer_step_total=61 lr=1.000e-05 loss=3.3535 vloss=0.0000 ploss=3.3535
(SpecoWorker pid=731468) WARNING:2026-09-02 18:17:48,270:[drafter loss] step=20 optimizer_step_total=80 lr=1.000e-05 loss=3.0168 vloss=0.0000 ploss=3.0168 [repeated 31x across cluster]
(WorkerDict pid=728170) WARNING:2026-09-02 18:18:00,813:[speco sglang draft update] done global_steps=8 buckets=3 bucket_mb=512 elapsed_sec=0.659
Training Progress: 100%|██████████| 8/8 [07:04<00:00, 50.26s/it]
Training Progress: 100%|██████████| 8/8 [07:04<00:00, 53.03s/it]
step:8 - drafter/collected_samples:np.float64(8.0) - drafter/trained:np.float64(1.0) - drafter/published:np.float64(1.0) - critic/score/mean:0.875 - response_length/mean:482.78125 - timing_s/gen:21.00071443617344
(WorkerDict pid=728175) WARNING:2026-09-02 18:18:00,946:[speco sglang draft update] done global_steps=8 buckets=3 bucket_mb=512 elapsed_sec=0.714 [repeated 7x across cluster]
EXIT_CODE=0
```

</details>

**Comment by crp0128** [comment]
Two issues:

1. Draft weights may be sent to the target worker. `DFlashWorkerV2` does not implement `update_weights_from_tensor()`, so the draft model may remain stale. Please explicitly update `draft_model_runner` and add a weight-change test.

2. `fields(ServerArgs)` is incompatible with current SGLang main because `ServerArgs` is not a standard dataclass. Please use SGLang’s `record_fields(ServerArgs)` helper.

Also, please rebase onto the latest `main` and preserve the draft-update lifecycle introduced by PR #81.

**Comment by khazic** [comment]
Thanks for the careful read, both points were real.

1. Draft weights reaching the target: confirmed on sglang main. The scheduler routes any update_weights_from_tensor without disable_draft_model to the spec worker, and DFlashWorkerV2 (DSparkWorkerV2 too) has no such method, so __getattr__ forwarded the draft publish to the target runner. c2c29c7 extends the existing eagle_update_weights patch to those workers: they get an EAGLEWorkerV2-style both-runners default under the same routing wrapper, so a draft publish now lands only on draft_model_runner and a target sync only on the target. Putting it in the patch registry matters because the scheduler is a spawned process; the registry patches are re-applied there. The runner call also falls back to ModelRunner.weight_updater, where main moved update_weights_from_tensor. New tests in test_sglang_spec_worker_weight_update_contract.py cover the weight change on both workers (draft changes, target untouched, and the reverse), the unrouted default, and the failure paths.

2. fields(ServerArgs): switched to sglang's record_fields with struct/dataclass fallbacks for older builds. This also fixes the custom_weight_loader probe, which read __dataclass_fields__ and answered False on main, so the draft publish had lost its route marker.

The branch is merged up to current main with the PR #81 lifecycle intact (vLLM keeps native_draft_update, SGLang keeps its own attach path). CI is not yet green on my side for sglang main itself since it needs a GPU run; happy to re-validate on a node if you want numbers.
